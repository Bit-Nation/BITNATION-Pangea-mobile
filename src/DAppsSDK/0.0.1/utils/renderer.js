// @flow

import React from 'react';
import components from '../components';

export type ComponentsJSON = { type: string, props: Object, children: Array<Object> };

/**
 * @desc Function to get element class or function from text representation, e.g. 'Text' means Text component.
 * @param {string} typeName String representation of the type.
 * @return {*} Class or function of component or null if type is unknown.
 */
export const getTypeElementFromText = (typeName: string) => components[typeName];

/**
 * @desc Validates props against list of valid names and filter only whitelisted ones.
 * Used to prevent DApp developer to pass something bad to components we provide.
 * @param {Object} props Props to validate.
 * @param {string[]} validProps Object containing arrays by keys `native`, `stateBased` and `callbacks` according to corresponding type of props.
 * @param {string} type String representation of component type.
 * @return {Object} Props filtered by groups.
 */
export const validateProps = (
  props: Object,
  validProps: { native: Array<string>, custom: Array<string>, callbacks: Array<string> },
  type: string,
) => {
  const nativeProps = {};
  const customProps = {};
  const callbackProps = {};
  Object.keys(props).forEach((propName) => {
    if (typeof propName !== 'string') {
      console.warn('Prop name is not a string');
      return;
    }

    if (validProps.native.includes(propName)) {
      // It's an allowed native prop.
      nativeProps[propName] = props[propName];
    } else if (validProps.custom.includes(propName)) {
      // It's an allowed custom prop.
      customProps[propName] = props[propName];
    } else if (propName.endsWith('ID') && validProps.callbacks.includes(propName.substring(0, propName.length - 2))) {
      // It's an allowed prop to perform a callback. That properties are marked by ID suffix.
      callbackProps[propName] = props[propName];
    } else {
      console.warn(`Prop ${propName} is not allowed on component ${type}`);
    }
  });

  return { nativeProps, customProps, callbackProps };
};

/**
 * @desc Renders JSON provided by DApp developer into react components.
 * @param {ComponentsJSON} json JSON object that represents components to render.
 * @param {?string} key Key of component (used for updates).
 * @param {Function} customPropsProvider Function that takes component class or function and props and returns new props if it's needed to add ones.
 * @param {*} parent Function or class of component that is parent of currently used or undefined for root.
 * @return {*} Tree of components that is ready to be render.
 */
export const renderJSON = (json: ComponentsJSON, key: ?string, customPropsProvider: (component: any, ownProps: Object) => Object, parent: any) => {
  const { type, children } = json;

  // This is the case when we have string literal inside children.
  // It's only allowed for Text.
  if (type === undefined && typeof (json) === 'string') {
    if (parent === components.Text) {
      return json;
    }

    console.warn(`String literal may only appear inside Text component. Found "${json}" outside.`);
    return null;
  }

  const component = components[type];

  if (component == null) {
    console.warn(`Trying to render unknown component type ${type}.`);
    return null;
  }

  if (component.validNativeProps != null && Array.isArray(component.validNativeProps) === false) {
    console.warn(`validNativeProps field of component ${type} should be an array of strings.`);
    return null;
  }
  if (component.customProps != null && Array.isArray(component.customProps) === false) {
    console.warn(`customProps field of component ${type} should be an array of strings.`);
    return null;
  }
  if (component.callbackProps != null && Array.isArray(component.callbackProps) === false) {
    console.warn(`callbackProps field of component ${type} should be an array of strings.`);
    return null;
  }

  const { props } = json;

  const { nativeProps = props, customProps = {}, callbackProps = {} } =
    validateProps(json.props, {
      native: component.validNativeProps || [],
      custom: component.customProps || [],
      callbacks: component.callbackProps || [],
    }, type);

  return React.createElement(
    component,
    {
      nativeProps,
      ...customProps,
      ...customPropsProvider(component, { ...nativeProps, ...customProps, ...callbackProps }),
      key,
    },
    children
      ? children.map((child, index) => renderJSON(child, `${index}`, customPropsProvider, component))
      : null,
  );
};
