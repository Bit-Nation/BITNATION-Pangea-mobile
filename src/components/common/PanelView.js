// @flow

import * as React from 'react';
import { View, Text } from 'react-native';
import { MediaQueryStyleSheet } from 'react-native-responsive';
import _ from 'lodash';

import GlobalStyles from '../../global/Styles';
import Button from './Button';

type Props = {
  /**
   * @desc Title of panel
   */
  title?: string,
  /**
   * @desc Style object of the title
   */
  titleStyle?: any,
  /**
   * @desc Icon of panel
   * @todo Fix icon to be the icon. Currently it's just a text.
   */
  icon?: string,
  /**
   * @desc Body text of panel
   */
  body?: string,
  /**
   * @desc Renders content between panel body and panel button
   */
  renderAdditionalInfo?: () => React.Node,
  /**
   * @desc Callback on panel button click. Button appears only if this one is passed.
   */
  onButtonClick?: () => void,
  /**
   * @desc Title of panel button.
   */
  buttonTitle?: string,
  /**
   * @desc Renders content below panel button.
   */
  renderBottom?: () => React.Node,
  /**
   * @desc Style object to be passed into children container.
   */
  childrenContainerStyle?: any,
  /**
   * @desc Style object to be applied on root view.
   */
  style?: any,
  /**
   * @desc Children to be rendered inside panel view.
   */
  children?: React.Node
};

const styles = MediaQueryStyleSheet.create({
  ...GlobalStyles,
});

/**
 * @desc Component that renders common panel view. It is used to provide list or mosaic layout.
 * @type {React.Component}
 */
export default class PanelView extends React.Component<Props> {
  static defaultProps: Object;

  renderHeader() {
    const { title, icon, titleStyle } = this.props;

    return (
      <View style={styles.panelTitleRowContainer}>
        {title && (
          <View style={styles.panelTitleContainer}>
            <Text style={titleStyle}>{title}</Text>
          </View>
        )}
        {icon && (
          <View style={styles.panelTitleIcon}>
            <Text style={styles.panelIcon}>{icon}</Text>
          </View>
        )}
      </View>
    );
  }

  render() {
    const {
      style,
      childrenContainerStyle,
      renderBottom,
      renderAdditionalInfo,
      children,
      title,
      icon,
      body,
      buttonTitle,
      onButtonClick,
    } = this.props;

    return (
      <View style={style}>
        {/* TITLE + ICON */}
        {/* Hide this view if no title or icon to avoid line below it. */}
        {(!_.isEmpty(title) || !_.isEmpty(icon)) && this.renderHeader()}

        {/* CHILDREN (MAIN) DISPLAY AREA */}
        <View style={[styles.panelChildrenContainer, childrenContainerStyle]}>
          {children}
        </View>

        {body && (
          <View style={styles.panelBodyContainer}>
            <Text style={styles.body}>{body}</Text>
          </View>
        )}

        <View style={styles.messageAdditionalInfoContainer}>
          {renderAdditionalInfo && renderAdditionalInfo()}
        </View>

        {onButtonClick && (
          <Button
            style={styles.panelButton}
            title={buttonTitle}
            onPress={onButtonClick}
          />
        )}

        <View style={styles.messageBottomContainer}>
          {renderBottom && renderBottom()}
        </View>
      </View>
    );
  }
}

PanelView.defaultProps = {
  style: styles.panelView,
  titleStyle: styles.panelViewTitle,
  children: null,
  childrenContainerStyle: {},
  renderBottom: null,
  buttonTitle: 'Button',
  onButtonClick: null,
  renderAdditionalInfo: null,
  body: undefined,
  icon: undefined,
  title: undefined,
};
