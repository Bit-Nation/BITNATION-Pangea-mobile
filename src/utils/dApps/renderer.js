// @flow

import React from 'react';

import Text from '../../components/dApps/Text';
import View from '../../components/dApps/View';
import TextInput from '../../components/dApps/TextInput';
import Button from '../../components/dApps/Button';

export type ComponentsJSON = { type: string, props: Object, children: Array<Object> };

/**
 * @desc Function to get element class or function from text representation, e.g. 'Text' means Text component.
 * @param {string} typeName String representation of the type.
 * @return {*} Class or function of component or null if type is unknown.
 */
export const getTypeElementFromText = (typeName: string) => {
  switch (typeName) {
    case 'Text':
      return Text;
    case 'View':
      return View;
    case 'TextInput':
      return TextInput;
    case 'Button':
      return Button;
    default:
      console.warn(`Unknown component type ${typeName}.`);
      return null;
  }
};

/**
 * @desc Validates props against list of valid names and filter only whitelisted ones.
 * Used to prevent DApp developer to pass something bad to components we provide.
 * @param {Object} props Props to validate.
 * @param {string[]} validProps Object containing arrays by keys `native`, `stateBased` and `callbacks` according to corresponding type of props.
 * @param {string} type String representation of component type.
 * @return {Object} Props filtered by groups.
 */
export const validateProps = (
  props: Object,
  validProps: { native: Array<string>, stateBased: Array<string>, callbacks: Array<string> },
  type: string,
) => {
  const nativeProps = {};
  const stateBasedProps = {};
  const callbackProps = {};
  Object.keys(props).forEach((propName) => {
    if (validProps.native.includes(propName)) {
      // It's an allowed native prop.
      nativeProps[propName] = props[propName];
    } else if (propName.endsWith('Path') && validProps.stateBased.includes(propName.substring(0, propName.length - 4))) {
      // It's an allowed prop to set or get root component state. That properties are marked by Path suffix.
      stateBasedProps[propName] = props[propName];
    } else if (propName.endsWith('ID') && validProps.callbacks.includes(propName.substring(0, propName.length - 2))) {
      // It's an allowed prop to perform a callback. That properties are marked by ID suffix.
      callbackProps[propName] = props[propName];
    } else {
      console.warn(`Prop ${propName} is not allowed on component ${type}`);
    }
  });

  return { nativeProps, stateBasedProps, callbackProps };
};

/**
 * @desc Renders JSON provided by DApp developer into react components.
 * @param {ComponentsJSON} json JSON object that represents components to render.
 * @param {?string} key Key of component (used for updates).
 * @param {Function} customPropsProvider Function that takes component class or function and props and returns new props if it's needed to add ones.
 * @param {*} parent Function or class of component that is parent of currently used or undefined for root.
 * @return {*} Tree of components that is ready to be render.
 */
export const renderJSON = (json: ComponentsJSON, key: ?string, customPropsProvider: (component: any, ownProps: Object) => Object, parent: any) => {
  const { type, children } = json;

  // This is the case when we have string literal inside children.
  // It's only allowed for Text.
  if (type === undefined && typeof (json) === 'string') {
    if (parent === Text) {
      return json;
    }

    console.warn(`String literal may only appear inside Text component. Found "${json}" outside.`);
    return null;
  }

  const component = getTypeElementFromText(type);

  if (component === null) {
    return null;
  }

  const { props } = json;

  const { nativeProps = props, stateBasedProps = {}, callbackProps = {} } =
    component.validNativeProps !== undefined
      ? validateProps(json.props, {
        native: component.validNativeProps,
        stateBased: Object.keys(component.stateBasedProps || {}),
        callbacks: component.callbackProps || [],
      }, type)
      : {};

  return React.createElement(
    component,
    { ...nativeProps, ...customPropsProvider(component, { ...nativeProps, ...stateBasedProps, ...callbackProps }), key },
    children
      ? children.map((child, index) => renderJSON(child, `${index}`, customPropsProvider, component))
      : null,
  );
};
