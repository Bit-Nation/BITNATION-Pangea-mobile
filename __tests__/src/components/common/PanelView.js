import React from 'react';
import renderer from 'react-test-renderer';

import PanelView from '../../../../src/components/common/PanelView';

describe('PanelView rendering', () => {

  test('Panel view with title and button', () => {
    const tree = renderer
      .create(<PanelView title='Test panel title' onButtonClick={() => null} buttonTitle='Test panel button'/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Panel view with title and body', () => {
    const tree = renderer
      .create(<PanelView title='Test panel title'
                         body='Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum'/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Panel view with title, body and button', () => {
    const tree = renderer
      .create(<PanelView title='Test panel title'
                         body='Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum'
                         onButtonClick={() => null}
                         buttonTitle='Test panel button'/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

});
