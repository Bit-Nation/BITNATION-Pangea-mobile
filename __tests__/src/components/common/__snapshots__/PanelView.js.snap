// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PanelView rendering Panel view with title and body 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "rgba(74,144,226,0.2)",
        "borderRadius": 8,
        "flexDirection": "column",
        "flexShrink": 1,
        "justifyContent": "space-between",
        "marginBottom": 8,
        "marginLeft": 0,
        "marginRight": 0,
        "marginTop": 8,
        "paddingBottom": 15,
        "paddingLeft": 15,
        "paddingRight": 15,
        "paddingTop": 12,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "backgroundColor": "transparent",
        "borderBottomColor": "#607D8B",
        "borderBottomWidth": 1,
        "borderStyle": "solid",
        "flexDirection": "row",
        "paddingBottom": 4,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "transparent",
          "flex": 4,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#FFFFFF",
            "fontFamily": "Roboto",
            "fontSize": 22,
            "fontWeight": "bold",
            "letterSpacing": 0,
            "lineHeight": 28,
            "textAlign": "left",
          }
        }
      >
        Test panel title
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "marginBottom": 8,
      }
    }
  />
  <View
    style={
      Object {
        "marginBottom": 8,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#72A4DE",
          "fontFamily": "Roboto",
          "fontSize": 17,
          "fontWeight": "normal",
          "letterSpacing": 0,
          "lineHeight": 22,
        }
      }
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum
    </Text>
  </View>
  <View
    style={Object {}}
  />
  <View
    style={Object {}}
  />
</View>
`;

exports[`PanelView rendering Panel view with title and button 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "rgba(74,144,226,0.2)",
        "borderRadius": 8,
        "flexDirection": "column",
        "flexShrink": 1,
        "justifyContent": "space-between",
        "marginBottom": 8,
        "marginLeft": 0,
        "marginRight": 0,
        "marginTop": 8,
        "paddingBottom": 15,
        "paddingLeft": 15,
        "paddingRight": 15,
        "paddingTop": 12,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "backgroundColor": "transparent",
        "borderBottomColor": "#607D8B",
        "borderBottomWidth": 1,
        "borderStyle": "solid",
        "flexDirection": "row",
        "paddingBottom": 4,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "transparent",
          "flex": 4,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#FFFFFF",
            "fontFamily": "Roboto",
            "fontSize": 22,
            "fontWeight": "bold",
            "letterSpacing": 0,
            "lineHeight": 28,
            "textAlign": "left",
          }
        }
      >
        Test panel title
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "marginBottom": 8,
      }
    }
  />
  <View
    style={Object {}}
  />
  <View
    style={
      Array [
        Object {
          "borderRadius": 15,
          "height": 30,
          "justifyContent": "center",
        },
        Object {
          "backgroundColor": "#1C497E",
        },
        Object {
          "marginTop": 13,
        },
      ]
    }
    title="Test panel button"
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginLeft": 13,
          "marginRight": 13,
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        buttonTitle={true}
        disabledButtonTitle={false}
        ellipsizeMode="tail"
        style={
          Array [
            undefined,
            undefined,
            Object {
              "color": "#72A4DE",
              "fontSize": 14,
              "letterSpacing": -0.02,
              "lineHeight": 19,
              "textAlign": "center",
            },
            false,
            undefined,
          ]
        }
      >
        Test panel button
      </Text>
    </View>
  </View>
  <View
    style={Object {}}
  />
</View>
`;

exports[`PanelView rendering Panel view with title, body and button 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "rgba(74,144,226,0.2)",
        "borderRadius": 8,
        "flexDirection": "column",
        "flexShrink": 1,
        "justifyContent": "space-between",
        "marginBottom": 8,
        "marginLeft": 0,
        "marginRight": 0,
        "marginTop": 8,
        "paddingBottom": 15,
        "paddingLeft": 15,
        "paddingRight": 15,
        "paddingTop": 12,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "backgroundColor": "transparent",
        "borderBottomColor": "#607D8B",
        "borderBottomWidth": 1,
        "borderStyle": "solid",
        "flexDirection": "row",
        "paddingBottom": 4,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "transparent",
          "flex": 4,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#FFFFFF",
            "fontFamily": "Roboto",
            "fontSize": 22,
            "fontWeight": "bold",
            "letterSpacing": 0,
            "lineHeight": 28,
            "textAlign": "left",
          }
        }
      >
        Test panel title
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "marginBottom": 8,
      }
    }
  />
  <View
    style={
      Object {
        "marginBottom": 8,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#72A4DE",
          "fontFamily": "Roboto",
          "fontSize": 17,
          "fontWeight": "normal",
          "letterSpacing": 0,
          "lineHeight": 22,
        }
      }
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum
    </Text>
  </View>
  <View
    style={Object {}}
  />
  <View
    style={
      Array [
        Object {
          "borderRadius": 15,
          "height": 30,
          "justifyContent": "center",
        },
        Object {
          "backgroundColor": "#1C497E",
        },
        Object {
          "marginTop": 13,
        },
      ]
    }
    title="Test panel button"
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginLeft": 13,
          "marginRight": 13,
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        buttonTitle={true}
        disabledButtonTitle={false}
        ellipsizeMode="tail"
        style={
          Array [
            undefined,
            undefined,
            Object {
              "color": "#72A4DE",
              "fontSize": 14,
              "letterSpacing": -0.02,
              "lineHeight": 19,
              "textAlign": "center",
            },
            false,
            undefined,
          ]
        }
      >
        Test panel button
      </Text>
    </View>
  </View>
  <View
    style={Object {}}
  />
</View>
`;
